libyang/xpath.py

```diff
@@ -56,17 +58,28 @@ def xpath_split(xpath: str) -> Iterator[Tuple[str, str, List[Tuple[str, str]]]]:
         while i < len(xpath) and xpath[i] == "[":
             i += 1  # skip opening '['
             j = xpath.find("=", i)  # find key name end
-            key_name = xpath[i:j]
-            quote = xpath[j + 1]  # record opening quote character
-            j = i = j + 2  # skip '=' and opening quote
-            while True:
-                if xpath[j] == quote and xpath[j - 1] != "\\":
-                    break
-                j += 1
-            # replace escaped chars by their non-escape version
-            key_value = xpath[i:j].replace(f"\\{quote}", f"{quote}")
-            keys.append((key_name, key_value))
-            i = j + 2  # skip closing quote and ']'
+
+            if j != -1:  # keyed specifier
+                key_name = xpath[i:j]
+                quote = xpath[j + 1]  # record opening quote character
+                j = i = j + 2  # skip '=' and opening quote
+                while True:
+                    if xpath[j] == quote and xpath[j - 1] != "\\":
+                        break
+                    j += 1
+                # replace escaped chars by their non-escape version
+                key_value = xpath[i:j].replace(f"\\{quote}", f"{quote}")
+                keys.append((key_name, key_value))
+                i = j + 2  # skip closing quote and ']'
+            else:  # index specifier
+                j = i
+                while True:
+                    if xpath[j] == "]":
+                        break
+                    j += 1
+                key_value = xpath[i:j]
+                keys.append((KEY_INDEX, key_value))
```

Comment:
I am a bit concerned that the `keys` list can now contain tuples that are not only strings. It would be less disruptive if the key name would be an empty string in this case.

You can check that below without much hassle.